L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,504 @@
+  FileWatcherCertificateProvider(const std::string& private_key_path,
+                                 const std::string& identity_certificate_path,
+                                 unsigned int refresh_interval_sec);
+
+  // Constructor to get credential updates from root file path only.
+  FileWatcherCertificateProvider(const std::string& root_cert_path,
+                                 unsigned int refresh_interval_sec);
+    }
+
+
+```
+### Customizable Verification of the Peer Certificate
+
+Users should be able to customize verification of the peer certificate, and we provide two distinct interfaces for how this can be done:
+
+1. Allow users to perform custom validation checks on the peer's certificate chain after it has been verified to chain up to a root of trust, but before the TLS handshake completes. This is a common customization that might be needed whenever a private PKI is involved, and does not require the user to have any significant domain knowledge.
+
+2. Allow users to fully customize path building and cryptographic verification. This is a relatively rare requirement from users and requires deep X.509 expertise to do correctly and without introducing significant security vulnerabilities. However, it is unavoidable in order to unblock some advanced use cases, e.g. SPIFFE federation.
+
+The next sections explain these interfaces in more detail.
+
+#### Custom Validation Checks on the Peer Certificate Chain
+```c++
+// Contains the information from the (verified) peer certificate chain that can
+// be used to perform custom validation checks. Users should not directly
+// create or destroy this request object.
+class TlsCustomVerificationCheckRequest {
+ public:
+  // The target hostname that is expected to appear in the server leaf
+  // certicate, e.g. github.com. Empty on the server-side.
+  absl::string_view target_name() const;
+
+  // The PEM-encoded leaf cert.
+  absl::string_view peer_cert() const;
+
+  // The PEM-encoded unverified peer cert chain, ordered from leaf to root, but
+  // not including the root.
+  absl::string_view peer_cert_full_chain() const;
+
+  // The common name string from the Subject extension in the peer's leaf
+  // certificate.
+  absl::string_view common_name() const;
+
+  // The subject name of the root certificate used to verify the peer chain
+  // If verification fails or the peer certificate is self-signed, this will be an
+  // empty string. If verification is successful, it is a comma-separated list,
+  // where the entries are of the form "FIELD_ABBREVIATION=string"
+  // ex: "CN=testca,O=Internet Widgits Pty Ltd,ST=Some-State,C=AU"
+  // ex: "CN=GTS Root R1,O=Google Trust Services LLC,C=US"
+  absl::string_view verified_root_cert_subject() const;
+
+  // The (possibly empty) list of URI names from the Subject Alternative Name
+  // extension in the peer's leaf certificate.
+  std::vector<absl::string_view> uri_names() const;
```

Comment:
From review, return absl::Span<absl::string_view> instead of std::vector<absl::string_view>.

Reply:
Similarly below.